numpy/core/src/multiarray/datetime.c

```diff
@@ -339,6 +361,15 @@ convert_datetimestruct_to_datetime(PyArray_DatetimeMetaData *meta,
                 break;
             case NPY_FR_fs:
                 /* only 2.6 hours */
+                if (abs(get_datetimestruct_seconds(dts)) > 
+                        2.6 * 60 * 60) {
+                    /* reject with seconds resolution */
+                    PyErr_SetString(PyExc_ValueError,
+                            "Time span around POSIX epoch is "
+                            "too large for fs units");
+                    return -1;
+                };
+
                 ret = (((((days * 24 +
```

Comment:
I don't know if this is a hot path. If it's not, can we use something like the following, which doesn't require us to hard-code these imprecise limits:
```C
ret = days;
err |= overflow_mult_then_add(&ret, 24, dts->hour);
err |= overflow_mult_then_add(&ret, 60, dts->min;
err |= overflow_mult_then_add(&ret, 60, dts->sec);
err |= overflow_mult_then_add(&ret, 1000000, dts->us);
err |= overflow_mult_then_add(&ret, 1000000, dts->ps);
err |= overflow_mult_then_add(&ret, 1000, dts->as / 1000);
```
Where
```C
bool overflow_mult_then_add(npy_int64 *val, npy_uint64 mult, npy_int64 add) {
    if (add > 0 && *val > (NPY_INT64_MAX - add) / mult) {
        return true;
    }
    else if (add < 0 && *val < (NPY_INT64_MIN - add) / mult) {
        return true;
    }
    else {
        *val = *val * mult + add;
        return false;
    }
}

Reply:
I assume that by "imprecise" you mean that the time spans from the epoch that are tolerated for given return time units could be specified at a higher precision? And avoiding "hard coding" by incrementally checking for overflow as the return value is propagated through the conversions above, instead of i.e., "+/- n seconds from the epoch are tolerated for these time units."?

Is this something we want in scope for this PR? If I do that, then wouldn't it be prudent to do it for all the return blocks & also add unit tests that probe the increased / improved precision?

We'd also likely want to update the docs to reflect improved precision / limits on time since epoch tolerated if we do that, and for docs we'd still want some kid of hard-"coded" numbers for users to go by, right? I suppose we could switch to "approximate" wording in docs, and just use the old values as rough guides & let users push those overflow limits as they see fit.

Reply:
I'm certainly happy to do the work (we'd need benchmarks, too?), so just making sure we've thought about this.

Reply:
What I mean is that this change will break existing code that uses values between the extremes you enforce here and the true extremes representable in the type.

Reply:
Regarding the limits - we can document the limit as `np.timedelta64(np.iinfo(np.int64).max(), 'ps') * 10^-12` etc

Reply:
What about moving the overflow checks into each of the functions named `get_datetimestruct_minutes()` and so on? Then each switch case would just call a function coded for its units, with each preceding function in the conversion chain also protected from overflow.